jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -45,43 +49,49 @@ def json_changebase(obj, changer):
 
 # Transaction serialization and deserialization
 
-def deserialize(tx):
-    if isinstance(tx, (str, unicode)) and re.match('^[0-9a-fA-F]*$', tx):
-        #tx = bytes(bytearray.fromhex(tx))
-        return json_changebase(
-            deserialize(binascii.unhexlify(tx)), lambda x: safe_hexlify(x))
-    # http://stackoverflow.com/questions/4851463/python-closure-write-to-variable-in-parent-scope
-    # Python's scoping rules are demented, requiring me to make pos an object
-    # so that it is call-by-reference
-    pos = [0]
-
-    #TODO: these are disgustingly ignorant of overrun errors!
+def deserialize(txinp):
+    if isinstance(txinp, basestring) and re.match('^[0-9a-fA-F]*$', txinp):
+        tx = BytesIO(binascii.unhexlify(txinp))
+        hexout = True
+    else:
+        tx = BytesIO(txinp)
+        hexout = False
+
+    def hex_string(scriptbytes, hexout):
+        if hexout:
+            return binascii.hexlify(scriptbytes)
+        else:
+            return scriptbytes
+
     def read_as_int(bytez):
-        pos[0] += bytez
-        return decode(tx[pos[0] - bytez:pos[0]][::-1], 256)
+        if bytez == 2:
+            return struct.unpack(b'<H', ser_read(tx, 2))[0]
+        elif bytez == 4:
+            return struct.unpack(b'<I', ser_read(tx, 4))[0]
+        elif bytez == 1:
+            return struct.unpack(b'B', ser_read(tx, 1))[0]
+        elif bytez == 8:
+            return struct.unpack(b'<Q', ser_read(tx, 8))[0]
+        else:
+            raise SerializationError('Asked to read unsupported %x bytes; bytez can only be 1, 2, 4 or 8' % bytez)
```

Comment:
Just a comment: raising here is appropriate since this is only called by `read_var_int` (where the calling argument bytez cannot be outside 1,2,4,8), and in a few other places where the calling value is a fixed integer.

(In other places in `deserialize` we have to anticipate attacker input, but not here).

Reply:
`ser_read` is designed to error on any overruns, where needed those errors can be caught

Reply:
do we have to put any try/catch statements somewhere to prevent crashes on invalid transactions, thrown by `ser_read`?

Reply:
not entirely sure there, depends on which functions need to accept untrusted input